### PR TITLE
3.next- Start implementing the Arguments class.

### PR DESCRIPTION
**Pull request by markstory**
Build out basic positional argument methods. I've skipped methods to fetch positional arguments by name, as I don't think it makes sense now. The whole point of positional arguments is that they are _positional_.

Refs #11137

### CODE REVIEW

File: src/Console/Arguments.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+     */
+    protected $options;
+
+    /**
+     * Constructor
+     *
+     * @param array $args Positional arguments
+     * @param array $options Named arguments
+     */
+    public function __construct(array $args, array $options)
```
Comment (lorenzo) [inline src/Console/Arguments.php]: Should it maybe have the ability to name arguments? like being able to address the 3rd argument by name.
I've found this to be very useful in other frameworks.

File: src/Console/Arguments.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+/**
+ * Provides an interface for interacting with
+ * a command's options and arguments.
+ */
+class Arguments
+{
+    /**
+     * Positional arguments.
+     *
+     * @var array
```
Comment (dereuromark) [inline src/Console/Arguments.php]: Are those actuall `string[]`?

File: src/Console/Arguments.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    public function __construct(array $args, array $options)
+    {
+        $this->args = $args;
+        $this->options = $options;
+    }
+
+    /**
+     * Get all positional arguments.
+     *
+     * @return array
```
Comment (dereuromark) [inline src/Console/Arguments.php]: Same here then.

File: src/Console/Arguments.php (hunk 1 above)
Comment (dereuromark) [reply src/Console/Arguments.php]: I would then use a similar approach of naming as I mentioned in the main ticket:

    getArgumentAt(index)
    hasArgumentAt(index)

and the new

    hasArgument(string)

maybe then.

Comment (markstory) [reply src/Console/Arguments.php]: @lorenzo Are you thinking of a map between names and indexes?

Comment (lorenzo) [reply src/Console/Arguments.php]: Yes, exactly

Comment (markstory) [reply src/Console/Arguments.php]: Ok. I was thinking of skipping that. But if you think it will be valuable to have. I'll add it in.

Comment (markstory) [reply src/Console/Arguments.php]: I've renamed the methods to include 'At'. I'll add the name based methods in a subsequent pull request.

**Comment by markstory** [comment]
@lorenzo @dereuromark I've added named argument support.

File: tests/TestCase/Console/ArgumentsTest.php [hunk 4]
```diff
@@ -57,9 +56,43 @@ public function testGetArgumentAt()
...
+
+    /**
+     * get arguments by name
+     *
+     * @return void
+     */
+    public function testGetArgument()
+    {
+        $values = ['big', 'brown', 'bear'];
+        $names = ['size' => 0, 'color' => 1, 'species' => 2, 'odd' => 3];
```
Comment (saeideng) [inline tests/TestCase/Console/ArgumentsTest.php]: how happen if  `$name` set like bellow 
```php
$names = ['size', 'color', 'species', 'odd'];
//or 
$names = [0 => 'size', 1 => 'color', 2 => 'species',3 =>  'odd'];
```

File: src/Console/Arguments.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+     * @param int $index The argument index to access.
+     * @return string|null The argument value or null
+     */
+    public function getArgumentAt($index)
+    {
+        if ($this->hasArgumentAt($index)) {
+            return $this->args[$index];
+        }
+
+        return null;
```
Comment (robertpustulka) [inline src/Console/Arguments.php]: Maybe the exception should be thrown if an argument is not present?

File: src/Console/Arguments.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+    /**
+     * Check if a positional argument exists by name
+     *
+     * @param string $name The argument name to check.
+     * @return string|null
+     */
+    public function getArgument($name)
+    {
+        if (!isset($this->argNames[$name])) {
+            return null;
```
Comment (robertpustulka) [inline src/Console/Arguments.php]: Same here.

File: src/Console/Arguments.php (hunk 5 above)
Comment (markstory) [reply src/Console/Arguments.php]: Is an argument missing exceptional? The CLI parser should throw errors when required arguments are missing.

File: tests/TestCase/Console/ArgumentsTest.php (hunk 4 above)
Comment (markstory) [reply tests/TestCase/Console/ArgumentsTest.php]: @saeideng We'd have to flip that data so we can use `isset()`, or switch to using `array_search()`. Using `array_search()` might be simpler for callers now that you mention it.

**Pull request merged**